### PR TITLE
Enable Multi GPU Training for AutoMM Detection

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
- Enabled multi gpu training.
- Added 2 options for validation metric: using val loss or mAP:
  - Using val loss is much faster and good for small models like `yolov3_mobilenetv2_320_300e_coco`
  - Using mAP gives better performance and the longer validation time is negligible for medium/large models.
- Copied torchmetrics mAP implementation and wait for their stable release with their issue 1147 and issue 1275 fixed.
- Completed TODOs:
  - infer `ROIS` type correctly in evaluation/prediction, and remove `col_name` hardcode
  - correctly transform label for detection and remove `reset_index`
  - move data unpack to collate function, remove `.to(self.device)` in `lit_mmdet`.

**For reviewers I'll suggest to skip the map.py where it contains the torchmetrics mAP implementation (in 0.8.0) with only one line fix as described in torchmetrics PR 1150).**

For the model's performance please see internal quip document.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/automm/object_detection/detection_train.py [hunk 1]
```diff
@@ -97,4 +109,7 @@ def detection_train(
         lr=args.lr,
         wd=args.wd,
         epochs=args.epochs,
+        num_gpus=args.num_gpus,
+        val_metric=args.val_metric,  # "mAP" or "direct_loss" or None (use default: "direct_loss")
```
Comment (sxjscience) [inline examples/automm/object_detection/detection_train.py:113]: I previously commented about using `mAP` instead of `map`. After a second thought, I think may be `map` is easier to remember.

File: multimodal/src/autogluon/multimodal/data/datamodule.py [hunk 2]
```diff
@@ -58,6 +59,10 @@ def __init__(
         id_mappings
              Id-to-content mappings. The contents can be text, image, etc.
              This is used when the dataframe contains the query/response indexes instead of their contents.
+        val_is_train
+             Whether we want to treat validation as training.
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/datamodule.py:63]: I think we should say `whether we are triggering `is_training` when creating the dataset for validation`. We can make the docstring more clear.

File: multimodal/src/autogluon/multimodal/data/datamodule.py [hunk 3]
```diff
@@ -58,6 +59,10 @@ def __init__(
...
+             This is used when we want to use val_loss as val metric, and thus training_pipeline
+             instead of val_pipeline during validation.
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/datamodule.py:65]: I think we can say `instead of the testing pipeline during validation`?

File: multimodal/src/autogluon/multimodal/data/process_image.py [hunk 4]
```diff
@@ -403,11 +413,11 @@ def process_one_sample(
         column_start = 0
         if self.prefix.lower().startswith(MMLAB_MODELS):
             for per_col_name, per_col_content in image_paths.items():
-                if per_col_name != "rois":  # TODO: remove hardcode
-                    mm_data["img_info"] = dict(filename=per_col_content[0])
-                else:
+                if is_rois_input(per_col_content):
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/process_image.py:416]: We should rely on the infer_type logic.

File: multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py [hunk 5]
```diff
@@ -219,29 +199,48 @@ def log_step_results(self, losses):
...
 
     def validation_epoch_end(self, validation_step_outputs):
-        # TODO: add mAP/mAR_per_class
-        mAPs = {"val_" + k: v for k, v in self.validation_metric.compute().items()}
-        mAPs["val_mAP"] = mAPs["val_map"]
-        self.print(mAPs)
-        self.log_dict(mAPs, sync_dist=True)
+        val_result = self.validation_metric.compute()
+        if self.use_loss:
+            self.print("val_loss: %.2f" % val_result.item())
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py:227]: I think we may always use `self.log()` and avoid `self.print()`. `self.print()` may only be used when debugging the code.

File: examples/automm/object_detection/detection_train.py (hunk 1 above)
Comment (FANGAreNotGnu) [reply examples/automm/object_detection/detection_train.py:113]: Sure, changing back to map, but keeping both `val_map` and `val_mAP` output keys in case users use `mAP` to retrieve results in their script.

File: multimodal/src/autogluon/multimodal/data/datamodule.py [hunk 6]
```diff
@@ -30,6 +30,7 @@ def __init__(
         test_data: Optional[pd.DataFrame] = None,
         predict_data: Optional[pd.DataFrame] = None,
         id_mappings: Optional[Dict[str, Dict]] = None,
+        val_is_train: bool = False,
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/datamodule.py]: Changed to val_use_training_mode.

File: multimodal/src/autogluon/multimodal/data/datamodule.py (hunk 2 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/datamodule.py]: Modified the docstring.

File: multimodal/src/autogluon/multimodal/data/datamodule.py (hunk 3 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/datamodule.py]: Modified the docstring.

File: multimodal/src/autogluon/multimodal/data/process_image.py (hunk 4 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/process_image.py:416]: The type inferred by `infer_type` is not forwarded to `process_image`, but here we use the same function that `infer_type` is using. Also we will move these code to `process_rois.py` later.

File: multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py (hunk 5 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py]: Done.

File: examples/automm/object_detection/detection_eval.py [hunk 7]
```diff
@@ -21,28 +21,36 @@
 
 
 def detection_evaluation(
-    checkpoint_name="yolov3_mobilenetv2_320_300e_coco", test_path="coco17/annotations/instances_val2017.json"
+    checkpoint_name="yolov3_mobilenetv2_320_300e_coco",
+    test_path="coco17/annotations/instances_val2017.json",
+    num_gpus=1,
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply examples/automm/object_detection/detection_eval.py:26]: Multigpu inference will be available in next PR.

File: examples/automm/object_detection/detection_eval.py [hunk 8]
```diff
@@ -21,28 +21,36 @@
...
         },
         pipeline="object_detection",
     )
 
     import time
 
     start = time.time()
-    predictor.evaluate(test_path)
+    result = predictor.evaluate(test_path)
     print("time usage: %.2f" % (time.time() - start))
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply examples/automm/object_detection/detection_eval.py:41]: In examples I think it's ok to use print? Changed to log in other files.

File: multimodal/src/autogluon/multimodal/data/infer_types.py [hunk 9]
```diff
@@ -112,16 +113,14 @@ def is_rois_column(data: pd.Series) -> bool:
     Whether the column is a rois column.
     """
     idx = data.first_valid_index()
-    if isinstance(data[idx], list):
-        return len(data[idx]) and isinstance(data[idx][0], list) and len(data[idx][0]) == 5
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/infer_types.py:116]: Added a TODO to better infer logic / input format to not confuse with other modality. Maybe we'll add more constraints (like check the format is [float, float, float, float, int]) / dictionary keys.

File: multimodal/src/autogluon/multimodal/data/utils.py [hunk 10]
```diff
@@ -386,3 +386,19 @@ def tokenize_ner_text(text, tokenizer):
     offset_mapping = np.array(col_tokens.offset_mapping, dtype=np.int32)
     word_offsets = np.pad(word_offsets, ((0, offset_mapping.shape[0] - len(words)), (0, 0)), "constant")
     return col_tokens, token_to_word_mappings, word_offsets
+
+
+def is_rois_input(sample):
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/utils.py:391]: Yes, added a TODO as above

File: multimodal/src/autogluon/multimodal/models/mmdet_image.py [hunk 11]
```diff
@@ -166,6 +178,9 @@ def forward(
         -------
             A dictionary with bounding boxes.
         """
+        # TODO: refactor this to work like forward() in MMDet
+        warnings.warn("MMDetAutoModelForObjectDetection.forward() is deprecated since it does not support multi gpu.")
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/models/mmdet_image.py]: Done

File: multimodal/src/autogluon/multimodal/utils/map.py [hunk 12]
```diff
@@ -0,0 +1,760 @@
+# Copyright The PyTorch Lightning team.
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/map.py:1]: I copied torchmetrics mAP implementation and wait for their stable release with https://github.com/Lightning-AI/metrics/issues/1147 and https://github.com/Lightning-AI/metrics/issues/1275 fixed.

File: multimodal/src/autogluon/multimodal/data/infer_types.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/data/infer_types.py:116]: Yes, we need a better logic. TODO is fine for now so that we can move fast.

**Comment by sxjscience** [comment]
@FANGAreNotGnu I think torchmetrics 0.10.1 has just been released https://github.com/Lightning-AI/metrics/releases/tag/v0.10.1 and it's available in pypi. We may not need `map.py` anymore.

**Review by sxjscience — APPROVED** [review]
LGTM overall. We should remove `map.py` and upgrade torchmetrics to be >=0.10.1.

**Review by bryanyzhu — APPROVED** [review]
LGTM. Let's fix the map.py in multigpu inference PR.

File: multimodal/src/autogluon/multimodal/models/mmdet_image.py [hunk 13]
```diff
@@ -166,6 +178,9 @@ def forward(
         -------
             A dictionary with bounding boxes.
         """
+        # TODO: refactor this to work like forward() in MMDet
+        logger.warning("MMDetAutoModelForObjectDetection.forward() is deprecated since it does not support multi gpu.")
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/mmdet_image.py:182]: I guess `realtime_predict` still uses this forward. Consider support realtime_predict later.

Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/models/mmdet_image.py:182]: Sure, added in the TODO here.
